concordium-consensus/src/Concordium/KonsensusV1/Types.hs

```diff
@@ -412,45 +444,73 @@ data FinalizationEntry = FinalizationEntry
     }
     deriving (Eq, Show)
 
-instance Serialize FinalizationEntry where
-    put FinalizationEntry{..} = do
-        put feFinalizedQuorumCertificate
-        let QuorumCertificate{..} = feSuccessorQuorumCertificate
-        put qcAggregateSignature
-        put qcSignatories
-        put feSuccessorProof
-    get = do
-        feFinalizedQuorumCertificate <- get
-        qcAggregateSignature <- get
-        qcSignatories <- get
-        feSuccessorProof <- get
-        let sqcRound = qcRound feFinalizedQuorumCertificate + 1
-        let sqcEpoch = qcEpoch feFinalizedQuorumCertificate
-        let feSuccessorQuorumCertificate =
-                QuorumCertificate
-                    { qcRound = sqcRound,
-                      qcEpoch = sqcEpoch,
-                      qcBlock =
-                        successorBlockHash
-                            ( BlockHeader
-                                sqcRound
-                                sqcEpoch
-                                (qcBlock feFinalizedQuorumCertificate)
-                            )
-                            feSuccessorProof,
-                      ..
-                    }
-        return FinalizationEntry{..}
-
-instance HashableTo Hash.Hash (Option FinalizationEntry) where
+-- | Convert a 'FinalizationEntry' to a 'ProtoFinalizationEntry'. This loses redundant information
+--  that can be recovered knowing the protocol version (if the finalization entry is well formed).
+toProtoFinalizationEntry :: FinalizationEntry pv -> ProtoFinalizationEntry
+toProtoFinalizationEntry FinalizationEntry{..} =
+    ProtoFinalizationEntry
+        { pfeFinalizedQuorumCertificate = feFinalizedQuorumCertificate,
+          pfeSuccessorAggregateSignature = qcAggregateSignature feSuccessorQuorumCertificate,
+          pfeSuccessorSignatories = qcSignatories feSuccessorQuorumCertificate,
+          pfeSuccessorProof = feSuccessorProof
+        }
+
+-- | Serialize a 'FinalizationEntry'. This is provided separately from the 'Serialize' instance
+--  since it does not depend on the @IsProtocolVersion pv@ constraint.
+putFinalizationEntry :: FinalizationEntry pv -> Put
+putFinalizationEntry = put . toProtoFinalizationEntry
+
+-- | Convert a 'ProtoFinalizationEntry' to a 'FinalizationEntry'. This depends on the protocol
+--  version for deriving the hash of the successor block from the 'ProtoFinalizationEntry' data.
+toFinalizationEntry' :: forall pv. SProtocolVersion pv -> ProtoFinalizationEntry -> FinalizationEntry pv
+toFinalizationEntry' spv ProtoFinalizationEntry{..} =
```

Comment:
Why do we need this (where the `'` in the name does not mean anything) as well as the `toFinalizationEntry` one?

Coudl we at least name them in a more informative way.

Reply:
Usually, having the `IsProtocolVersion` constraint it preferable, but in some cases we have an `SProtocolVersion` instead (e.g. with `getFinalizationEntry`). It isn't convenient to go from the latter to the former, so we provide both versions.